README.md
# Governify Render
<!-- [![Node.js CI](https://github.com/governify/render/workflows/Node.js%20CI/badge.svg?branch=master)](https://github.com/governify/render/actions)
[![Coverage Status](https://coveralls.io/repos/github/governify/render/badge.svg)](https://coveralls.io/github/governify/render) -->º
<a href="https://standardjs.com"><img src="https://img.shields.io/badge/code_style-semistandard-brightgreen.svg" alt="Standard - JavaScript Style Guide"></a>
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)

Please refer to  [Governify docs](https://docs.governify.io) to get the documentation.

### Running the server
To run the server, run:

```
npm start
```

## Copyright notice

Please refer to  [Governify License](https://docs.governify.io/about/license/) to get Copyright and License details.


CHANGELOG.md
## [2.3.1](https://github.com/governify/render/compare/v2.3.0...v2.3.1) (2023-11-08)



# [2.3.0](https://github.com/governify/render/compare/v2.2.0...v2.3.0) (2023-02-12)


### Bug Fixes

* NewRelic error handling ([ec860d2](https://github.com/governify/render/commit/ec860d2647d36ffc9620fde22030ee8020ba56fe))


### Features

* Installed newrelic ([7e9be37](https://github.com/governify/render/commit/7e9be37417accf6644995ab3849600e45a6ffa0c))



# [2.2.0](https://github.com/governify/render/compare/v2.1.0...v2.2.0) (2021-10-28)


### Bug Fixes

* body parser before commons middleware ([f1238c4](https://github.com/governify/render/commit/f1238c4b1eca97f08996b401346f3bafefe0b35a))
* github workflow ([14a75fa](https://github.com/governify/render/commit/14a75fa45f63aacebc58dbe1ed52ce2b46d4c30f))
* rm collectors from external infrastructure ([45812dd](https://github.com/governify/render/commit/45812dda2ce6564f2de1086349236da654d0db47))


### Features

* New logs implementation ([3624262](https://github.com/governify/render/commit/36242627a3d24a79214a66f1d06288d9302233d2))
* update commons v1.14 ([1afa966](https://github.com/governify/render/commit/1afa96605abf56131ff69a06a6b82238f3f31b9c))
* updated bootstrap5 ([3eda96c](https://github.com/governify/render/commit/3eda96caf4a1328d9b01d3ad8c9deef0fed5ce69))



# [2.1.0](https://github.com/governify/render/compare/32077a5126b26a3824eb578c98ca89de9ac09ea2...v2.1.0) (2021-05-07)


### Bug Fixes

* databases in infrastructure ([7c99238](https://github.com/governify/render/commit/7c99238b3cc1b75b076bce1f29bab7464cb28075))
* Default envvars ([f52b8ac](https://github.com/governify/render/commit/f52b8ac43a28cbc4e22bfb2cf5f553af6ed84778))
* removed linting and testing from workflow ([867034b](https://github.com/governify/render/commit/867034b8bebb70704c25b5cfb54cacc625d107c8))
* URL escapes for mustache ([abd0211](https://github.com/governify/render/commit/abd0211ab48fb9b544ec052152c9d1295ce4a23d))


### Features

* commons middleware ([bc654b5](https://github.com/governify/render/commit/bc654b5f3d5d9da3a9f01ed7812190b95a5840b3))
* Dockerfile update to match with the architecture ([f1b65ea](https://github.com/governify/render/commit/f1b65eae56a892cd5c2f6f7f09b94b27d8098a63))
* First version of render 2.0 ([32077a5](https://github.com/governify/render/commit/32077a5126b26a3824eb578c98ca89de9ac09ea2))
* lint, test, CI/CD, cleaning up ([3373a84](https://github.com/governify/render/commit/3373a84616f96709a88c77702beeeaca96319f03))
* new docker image in use ([7fcacfc](https://github.com/governify/render/commit/7fcacfcf5188f7c28ec71897cb9ac3e07449dd57))
* package author, repository and docker ([b4715b7](https://github.com/governify/render/commit/b4715b77a7ecd5ab5a140f71ad352779c9ed2409))
* release, build and push ([b5e0a71](https://github.com/governify/render/commit/b5e0a715fc7fd50e377c0ca5835145a2e3d96c59))
* replaced requests and port GC ([c5772e2](https://github.com/governify/render/commit/c5772e2bc7755d3b6c1e9a12be26a77cb0da5676))
* update commons ([127b471](https://github.com/governify/render/commit/127b471feaecc46eccbfe82230191ff8b05f2cd5))
* updated governify-commons ([88f5f06](https://github.com/governify/render/commit/88f5f062a82ad4e3910206d255fb448327619591))
* Updating to new moustache variables ([535be1a](https://github.com/governify/render/commit/535be1a2166646b788cc67137008f082d2a59ce3))





index.js
'use strict';

require('newrelic');

const governify = require('governify-commons');
const logger = governify.getLogger().tag('initialization');

const server = require('./server');

const env = process.env.NODE_ENV ? process.env.NODE_ENV : 'production';

governify.init().then((commonsMiddleware) => {
  server.deploy(env, commonsMiddleware).catch(err => { logger.error(err); });
});

// quit on ctrl-c when running docker in terminal
process.on('SIGINT', function onSigint () {
  logger.info('Got SIGINT (aka ctrl-c in docker). Graceful shutdown ', new Date().toISOString());
  shutdown();
});

// quit properly on docker stop
process.on('SIGTERM', function onSigterm () {
  logger.info('Got SIGTERM (docker container stop). Graceful shutdown ', new Date().toISOString());
  shutdown();
});

const shutdown = () => {
  server.undeploy();
};


server.js
/*!
governify-render 1.0.0, built on: 2018-05-09
Copyright (C) 2018 ISA group
http://www.isa.us.es/
https://github.com/isa-group/governify-render#readme

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>. */

'use strict';




function check (name, pass, config) {
  const nameOK = config.auth.user;
  const passOK = config.auth.password;

  let valid = true;
  valid = name === nameOK && valid;
  valid = pass === passOK && valid;
  return valid;
}

const deploy = (env, commonsMiddleware, callback) => {
  return new Promise((resolve, reject) => {
    try {
      const bodyParser = require('body-parser');
      const express = require('express');
      const cors = require('cors');
      const helmet = require('helmet');
      const compression = require('compression');
      const basicAuth = require('basic-auth');
      const path = require('path');
      const governify = require('governify-commons');
      const logger = governify.getLogger().tag('initialization');
      const config = require('./src/backend/configurations');
      

      const app = express();
      app.use(commonsMiddleware);

      if (config.server.enableHTTPBasicAuth) {
        logger.info("Adding 'WWW-Authenticate:' header to every path. Check config/env for getting user and pass");
        app.use((req, res, next) => {
          const credentials = basicAuth(req);
          if (!credentials || !check(credentials.name, credentials.pass, config)) {
            res.statusCode = 401;
            res.setHeader('WWW-Authenticate', 'Basic realm="Snapshot Management Login"');
            res.end('Unauthorized');
          } else {
            next();
          }
        });
      }

      const routes = require('./src/backend/routes.js');

      app.use(compression());

      app.use(
        bodyParser.urlencoded({
          limit: '50mb',
          extended: 'true'
        })
      );

      app.use(
        bodyParser.json({
          limit: '50mb',
          type: 'application/json'
        })
      );

      const frontendPath = path.join(__dirname, '/src/frontend');
      logger.info("Serving '%s' as static folder", frontendPath);
      app.use(express.static(frontendPath));

      if (config.server.bypassCORS) {
        logger.info("Adding 'Access-Control-Allow-Origin: *' header to every path.");
        app.use(cors());
      }
      if (config.server.useHelmet) {
        logger.info('Adding Helmet related headers.');
        app.use(helmet());
      }

      // Bypassing 405 status put by swagger when no request handler is defined
      app.options('/*', (req, res, next) => {
        return res.sendStatus(200);
      });

      const serverPort = process.env.PORT || config.server.port;

      app.listen(serverPort, () => {
        logger.info('Your server is listening on port %d (http://localhost:%d)', serverPort, serverPort);
      });

      // For all GET requests, send back index.html
      app.use('/', routes);

      // app.use('/', express.static(__dirname + '/public'));
    } catch (err) {
      reject(err);
    }
  });
};

const undeploy = () => {
  process.exit();
};

module.exports = {
  deploy: deploy,
  undeploy: undeploy
};


src/backend/routes.js
/*!
governify-render 1.0.0, built on: 2018-05-09
Copyright (C) 2018 ISA group
http://www.isa.us.es/
https://github.com/isa-group/governify-render#readme

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>. */

'use strict';

const express = require('express');
const router = express.Router();
const path = require('path');
const fs = require('fs');
const jsyaml = require('js-yaml');
const mustache = require('mustache');
mustache.escape = function (text) { return text; };
const governify = require('governify-commons');

const config = require('./configurations');
const logger = governify.getLogger().tag("routes")

module.exports = router;

router.get('/', function (req, res) {
  var viewPath = config.default_render.view;
  var ctrlPath = config.default_render.controller;
  var modelPath = config.default_render.model;
  logger.info('Redirecting to /render?model=' + modelPath + '&view=' + viewPath + '&ctrl=' + ctrlPath);
  res.redirect('/render?model=' + modelPath + '&view=' + viewPath + '&ctrl=' + ctrlPath);
});

async function getFileFromString (URL) {
  var file;
  if (URL.startsWith('https://') || URL.startsWith('http://')) {
    logger.info('Getting external file: ' + URL);
    await governify.httpClient({
      url: URL,
      method: 'GET',
      headers: { 'User-Agent': 'request' },
      transformResponse: function (response) {
        // do not convert the response to JSON or object
        return response;
      }
    }).then(response => {
      file = response.data;
    }).catch(err => {
      console.error(err);
      throw Error('Error obtaining: ' + URL);
    });
  } else {
    // Check first file exist to return null
    if (fs.existsSync('./src/frontend' + URL)) {
      logger.info('Getting file locally: ' + URL);
      file = await fs.readFileSync('./src/frontend' + URL, 'utf8');
    } else {
      throw Error('Error obtaining: ' + URL);
    }
  }
  // Compatibility for yaml files.
  if (URL.endsWith('.yaml')) {
    file = JSON.stringify(jsyaml.safeLoad(file));
  }
  return file;
}

router.get('/render', async function (req, res) {
  var ctrl = req.query.ctrl;
  var model = req.query.model;
  var view = req.query.view;

  if (!ctrl || !view || !model) {
    logger.warn('No params');
    res.sendStatus(404);
  } else {
    var files;
    try {
      const [fileCtrl, fileModel, fileView] = await Promise.all([getFileFromString(ctrl), getFileFromString(model), getFileFromString(view)]);
      files = {
        FILE_CONTROLLER: fileCtrl,
        FILE_MODEL: fileModel,
        FILE_VIEW: fileView
      };
      logger.info('Displaying render');
      var htmlTemplate = fs.readFileSync('./src/backend/layouts/' + config.layout, 'utf8');
      var htmlRendered = mustache.render(htmlTemplate, files, {}, ['$_[', ']']);
      res.send(htmlRendered);
    } catch (err) {
      logger.warn('Error getting files: ' + err);
      res.status(404).send('404 Not found - ' + err.message);
    }
  }
});

router.post('/updateDefaultTPA', function (req, res) {
  var tpaPath = path.join(__dirname, './../frontend/renders/tpa/template.json');
  fs.writeFileSync(tpaPath, JSON.stringify(req.body, null, 4));
  res.sendStatus(204);
});
